### ISSUE TITLE
Dropdown accessibility considerations

### ISSUE THREAD

**Issue by patrickhlauke** [twbs/bootstrap#18552]
meta-issue to discuss some of the points raised by http://www.slideshare.net/maxdesign/front-end-frameworks-are-they-accessible/45

> Problem 1: [...] there is often no assistance given to explain what they are being asked to do (such as choose from a range of options) when they arrive at the button

this will depend on specific use/context. if the dropdown is used for navigation, it should be fairly self-evident what the user can do. if it's more like application menus, titled in a similar way (e.g. "View", "Edit", "Help"), again it should make sense. It would be good to get a specific example in the wild of where this is not the case...but also, there's probably little advice we can give/example we can create beyond "make sure the trigger for your dropdown makes sense..." which is not something we can mandate at framework level.

> Problem 2: these elements are announced as buttons, so Screen Reader users often assume these are action orientated.

And they are - the action they perform is to open/close the dropdown menu?

> However, when these buttons are activated, Screen Reader users are then presented with silence. There is nothing to tell them that the dropdown has now visually opened...

Depending on screen reader, the change in `aria-expanded` from false to true will be communicated. Also, the presence of `aria-haspopup` should have already had some effect in the way the button was announced to indicate it has a menu.

However, the proposed solution of moving focus to the first item when the dropdown is opened is good...for keyboard users. For mouse users, it may look odd, and doesn't match known behavior of something like an actual application menu. May require different code paths for "opened with a mouse" vs "opened with keyboard", which gets ugly. May need to make a proof of concept to see how jarring it may be.

> Problem 3: for keyboard-only users, these dropdowns look similar to select menus, however the dropdown options cannot be accessed by standard select menu keystrokes - like the arrow keys

Although we're not strictly following this pattern, it may be worth adding functionality that opens the menu also for up/down arrow (a la http://www.w3.org/TR/wai-aria-practices/#menu). Once we go down this road, though, other similar keyboard interactions typicially associated with ARIA menus (like arrows to move between menu items etc) will probably also be expected ... but we shied aways from the pattern as dropdowns can be used for things that aren't just traditional ARIA menus.

I also wonder how prominent a problem this would be, and likely depends again on specific implementation/usage. if the dropdown is part of a navigation bar, users likely won't think it's a select. if it's added as part of a form though, they certainly may...hmmm

> Problem 4: when keyboard-only users activate the dropdown, then tab through the dropdown, the dropdown remains "open" even after they leave. These dropdowns only close via the button element.

Fair point, something we should address. However, once you try and just prevent user from TAB-ing out, what's to stop them using other means of jumping around the page (if they're AT users, e.g. using landmark navigation or similar)? Do we cater for that as well, or just the simple scenario of non-AT sighted keyboard user? any implications as well for sighted non-AT touchscreen users?

> Problem 5: [...] when users choose an item from the dropdown list, this choice is not reflected in the button [...]

So this is more to do with using dropdowns as faked selects? We don't encourage (or even show any examples of) this use, but I don't doubt that some people do (though I think it would need extra third party plugin/custom code to actually work)?

**Comment by cvrebert** [comment]
1. Yeah, not really a framework problem.
2. Yeah, my understanding was that our `aria-*` attributes should already take care of the screenreader presentation. As to moving the focus, I've no opinion.
3. I'll leave this to the experts. :wink: 
4. The slides propose merely closing the dropdown when the user tabs away from it (which I interpreted as "when the user focuses an element that's outside of the dropdown"), not trying to hold the focus hostage. That seems within the realm of reasonable.
5. Yeah, agreed. The most I could see here is a note like "If you're making this into a `<select>`, you're opening a can of worms which is beyond our scope. There Be Dragons, but here are some links: [...]"

**Comment by patrickhlauke** [comment]
Closing this, as in relation to 3. we specifically document now that dropdowns are not really ARIA-style menus, and for 4. our dropdowns now close once focus shifts away from them.
